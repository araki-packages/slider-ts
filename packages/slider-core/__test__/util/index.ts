
let oneFrameAtMS = 1000 / 120;
export const useSpyRequestAnimationFrame = () => {
  beforeEach(() => {
    let i = 0;
    jest.spyOn(window, 'requestAnimationFrame')
      .mockImplementation((callback: FrameRequestCallback) => {
        i += oneFrameAtMS;
        callback(i);
        return Math.random();
      });
  });
};

export const useSpyPerformanceNow = () => {
  beforeEach(() => {
    let i = 100;
    jest.spyOn(window.performance, 'now')
      .mockImplementation(() => {
        i += oneFrameAtMS;
        return i;
      });
  });
};

export const sleep = (time: number) => {
  return new Promise((r) => {
    setTimeout(() => {
      r();
    }, time);
  })
}