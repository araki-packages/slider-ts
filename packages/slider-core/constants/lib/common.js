"use strict";
exports.__esModule = true;
exports.InitialSliderOptions = void 0;
var cubic_bezier_1 = require("@takumus/cubic-bezier");
var EASE_OUT = [0, 0, 0.58, 1];
exports.InitialSliderOptions = {
    isLoop: true,
    wrapperWidth: 0,
    isFit: false,
    initialIndex: 0,
    smooth: 1,
    bezier: cubic_bezier_1["default"].apply(void 0, EASE_OUT)
};
