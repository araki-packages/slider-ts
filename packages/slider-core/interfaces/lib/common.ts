export interface SliderOptions {
  isLoop?: boolean; // 無限スクロールするかどうか
  isFit?: boolean; // 位置を同期させるかどうか
  initialIndex?: number; // indexの初期位置
  smooth?: number // 0 - 1 default 1
}
