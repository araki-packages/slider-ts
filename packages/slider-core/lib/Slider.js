"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
exports.__esModule = true;
exports.Slider = void 0;
var SpeedCalculator_1 = require("./SpeedCalculator");
var common_1 = require("../constants/lib/common");
var Slider = /** @class */ (function () {
    /**
     * カルーセルクラス
     * @param width
     * @param isLoop 無限ループさせるかどうか
     */
    function Slider(caches) {
        if (caches === void 0) { caches = 10; }
        this._position = 0; // 現在の位置情報
        this.elementNum = 0; // 総エレメント巣
        this.prevUpdateProps = {
            time: 0,
            position: 0
        };
        this.speedCalc = new SpeedCalculator_1.SpeedCalculator(caches);
    }
    Object.defineProperty(Slider.prototype, "position", {
        get: function () {
            return this._position;
        },
        set: function (position) {
            var nextPosition = this.updateLocation(position);
            this._position = nextPosition;
            this.handleChange(nextPosition);
        },
        enumerable: false,
        configurable: true
    });
    // 初期化イベント
    /**
     * @param width 描画範囲
     * @param viewElementNum 描画範囲内のアイテム数
     * @param options.isLoop
     * @param options.isFit
     * @param options.initialIndex
     * @param options.smooth - 移動時のなめらかさ
     */
    Slider.prototype.init = function (width, viewElementNum, option) {
        this.option = __assign(__assign({}, common_1.InitialSliderOptions), option);
        this.option.smooth = Math.max(this.option.smooth, 0.01);
        this.width = width;
        this.elementNum = viewElementNum;
        this.itemWidth = this.width / this.elementNum;
        this.position = this.option.initialIndex * this.itemWidth;
        this.bezierFn = this.option.bezier;
    };
    // タッチ or クリック 開始時に呼ばれるメソッド
    Slider.prototype.start = function (position) {
        cancelAnimationFrame(this.rafToken || 0);
        this.prevUpdateProps = {
            position: position,
            time: performance.now()
        };
        this.speedCalc.reset();
    };
    // 場所のアップデート
    Slider.prototype.update = function (position) {
        var _a = this.prevUpdateProps, pPosition = _a.position, pTime = _a.time;
        var moveOffset = pPosition - position;
        var prevTime = pTime - performance.now();
        this.position += moveOffset;
        this.speedCalc.add((moveOffset / prevTime) * 50);
        this.prevUpdateProps = {
            time: performance.now(),
            position: position
        };
    };
    // タッチイベント終了時
    Slider.prototype.end = function () {
        var speed = this.speedCalc.get() * 10 * this.option.smooth; // 重み付けの差
        var calcTime = Math.min(Math.max(Math.abs(speed), 100), 1000); // 速度の算出（最低２００ｍｓ）
        this.moveTo(speed, calcTime * this.option.smooth);
    };
    // 次のスライド
    Slider.prototype.next = function (duration) {
        if (duration === void 0) { duration = 100; }
        this.moveTo(this.itemWidth, duration);
    };
    // 前のスライド
    Slider.prototype.prev = function (duration) {
        if (duration === void 0) { duration = 100; }
        this.moveTo(-this.itemWidth, duration);
    };
    // TODO
    Slider.prototype.setIndex = function (index, duration) {
        if (duration === void 0) { duration = 1000; }
        var target = index * this.itemWidth;
        this.moveTo(target - this.position, duration);
    };
    // 慣性スクロール
    Slider.prototype.moveTo = function (movementPosition, maxTime) {
        var _this = this;
        // 最終到達地点が半分かそうじゃないかで分岐
        // ここが変
        var calcMovementPosition = this.option.isFit
            ? Math.floor(movementPosition -
                // koko
                (this.itemWidth - (this.position % this.itemWidth)) +
                // koko
                (this.itemWidth - (movementPosition % this.itemWidth)))
            : movementPosition;
        var startTime = null;
        var tick = function (time) {
            if (startTime === null)
                startTime = time;
            var elapsedTime = time - startTime;
            if (elapsedTime >= maxTime) {
                cancelAnimationFrame(_this.rafToken || 0);
                _this.position = _this.position - calcMovementPosition;
                _this.onEnd && _this.onEnd();
                return;
            }
            var elapsedParsentage = elapsedTime / maxTime;
            var bezierParsentage = _this.bezierFn(elapsedParsentage);
            _this.position = _this.position - bezierParsentage * calcMovementPosition;
            _this.rafToken = requestAnimationFrame(tick);
        };
        requestAnimationFrame(tick);
    };
    Slider.prototype.handleChange = function (nextPosition) {
        var index = Math.round(nextPosition / this.itemWidth);
        this.onChange && this.onChange(this.position, index);
    };
    // ロケーションのアップデート
    Slider.prototype.updateLocation = function (nextPosition) {
        if (this.option.isLoop) {
            return this.updateLocationIsLoop(nextPosition);
        }
        return this.updateLocationIsNotLoop(nextPosition);
    };
    Slider.prototype.updateLocationIsLoop = function (nextPosition) {
        var maxLength = this.itemWidth * (this.elementNum - 1);
        if (nextPosition < 0) {
            return maxLength + (this.position % maxLength);
        }
        return nextPosition % maxLength;
    };
    Slider.prototype.updateLocationIsNotLoop = function (nextPosition) {
        var maxLength = this.itemWidth * this.elementNum;
        var minPosition = 0;
        var maxPosition = maxLength - this.option.wrapperWidth;
        if (nextPosition < minPosition) {
            return 0;
        }
        if (nextPosition > maxPosition) {
            return maxLength - this.itemWidth - this.option.wrapperWidth;
        }
        return nextPosition;
    };
    return Slider;
}());
exports.Slider = Slider;
