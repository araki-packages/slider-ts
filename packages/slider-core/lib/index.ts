
console.log('hoge')
export type hoge = 'fuga' | 'bar';
export const hoge = (hoge: string) => {
  console.log(hoge);
}
