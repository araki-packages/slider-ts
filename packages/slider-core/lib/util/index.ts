export const deltaAnimation = (
  cb: (deltaTime: number, elapsedTime: number) => void
) => {
  let prevTime: number | null = null;
  let elapsedTime = 0;
  let id: null | number = null;
  const tick = (currentTime: number) => {
    id = window.requestAnimationFrame(tick);
    if (prevTime == null) prevTime = currentTime - 16;
    const deltaTime = -prevTime + currentTime;
    elapsedTime += deltaTime;
    cb(deltaTime, elapsedTime);
  };
  id = window.requestAnimationFrame(tick);

  return () => {
    if (id === null) return;
    window.cancelAnimationFrame(id);
  };
};
