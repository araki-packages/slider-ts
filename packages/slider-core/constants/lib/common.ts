import { SliderOptions } from '../../interfaces/lib/common';

export const InitialSliderOptions: Required<SliderOptions> = {
  isLoop: false,
  wrapperWidth: 0,
  isFit: false,
  initialIndex: 0,
  smooth: 1,
}