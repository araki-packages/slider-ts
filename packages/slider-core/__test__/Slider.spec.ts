import { Slider } from "../lib/Slider";
import { useSpyRequestAnimationFrame, useSpyPerformanceNow } from "./util";

class SliderLogger {
  private list: Array<{ x: number; i: number }> = [];

  public add(x: number, i: number): void {
    this.list.push({ x, i });
  }

  public clear(): void {
    this.list = [];
  }

  public toCSV(): string {
    const table = "'position', 'index'";
    const record = this.list.map(({ x, i }) => `${x}, ${i}`).join("\n");

    return `${table}\n${record}`;
  }
}

const itemWidth = 100;
const elementNum = 5;
const initialWidth = itemWidth * elementNum;
// const maxWidth = itemWidth * elementNum;
useSpyRequestAnimationFrame();
useSpyPerformanceNow();

describe("基本動作テスト", () => {
  it("Index初期位置ズレのテスト", () => {
    const instance = new Slider();
    const offsetIndex = 1;
    instance.onChange = (x, index) => {
      expect(index).toBe(1);
      expect(x).toBe(itemWidth * offsetIndex);
    };
    instance.init(initialWidth, elementNum, {
      initialIndex: offsetIndex,
      smooth: 0.1,
    });
  });

  it("ループのテスト:スナップショット", () => {
    const width = itemWidth * elementNum;
    const instance = new Slider();
    const logList = new SliderLogger();

    instance.onChange = (x, i) => {
      logList.add(x, i);
    };

    instance.init(width, elementNum, {
      isFit: true,
      isLoop: true,
      smooth: 0.1,
    });
    instance.onEnd = () => {
      expect(logList.toCSV()).toMatchSnapshot();
      logList.clear();
    };
    instance.start(5);
    instance.update(-100);
    instance.update(-400);
    instance.end();
    instance.start(5);
    instance.update(100);
    instance.update(400);
    instance.end();
  });

  it("非ループのテスト:スナップショット", () => {
    const instance = new Slider();
    let rx = 0;
    let ri = 0;
    const logList = new SliderLogger();

    instance.onChange = (x, i) => {
      logList.add(x, i);
      rx = x;
      ri = i;
    };
    instance.init(initialWidth, elementNum, {
      isFit: true,
      isLoop: false,
      smooth: 0.1,
    });
    instance.onEnd = () => {
      expect(rx).toBe(ri * itemWidth);
      expect(logList.toCSV()).toMatchSnapshot();
      logList.clear();
    };

    instance.start(5);
    instance.update(-50);
    instance.update(-100);
    instance.end();

    instance.start(5);
    instance.update(50);
    instance.update(100);
    instance.end();
  });

  it("限界値テスト", () => {
    const instance = new Slider();
    const logList = new SliderLogger();

    instance.onChange = (x, i) => {
      logList.add(x, i);
      expect(x).toBeGreaterThanOrEqual(0);
      //expect(x).toBeLessThanOrEqual(elementNum * itemWidth);
    };
    instance.init(initialWidth, elementNum, {
      isLoop: true,
      smooth: 0.1,
    });
    const updateLocation = (): void => {
      instance.start(5);
      instance.update(-100);
      instance.update(-400);
      instance.end();
      instance.start(5);
      instance.update(100);
      instance.update(400);
      instance.end();
    };
    updateLocation();
    instance.init(initialWidth, elementNum, {
      isLoop: false,
      smooth: 0.1,
    });
    updateLocation();
  });

  it("非ループ値限界値テスト", () => {
    const instance = new Slider();
    // let rx = 0;
    const logList: Array<{ x: number; i: number }> = [];

    instance.onChange = (x, i) => {
      logList.push({ x, i });
    //  rx = x;
    };
    instance.init(initialWidth, elementNum, {
      isLoop: false,
      smooth: 0.1,
    });

    // instance.onEnd = () => {
    //   expect(rx).toBe(0);
    // };
    instance.start(5);
    instance.update(25);
    instance.update(50);
    instance.update(75);
    instance.end();

    // instance.onEnd = () => {
    //   expect(rx).toBe(maxWidth - itemWidth);
    // };
    instance.start(0);
    instance.update(-25);
    instance.update(-50);
    instance.update(-75);
    instance.end();
  });

  it("インデックス移動テスト", () => {
    const instance = new Slider();
    instance.init(initialWidth, elementNum);

    const testIndex = (index: number): void => {
      // let ri = 0;
      // instance.onChange = (_, i) => {
      //   ri = i;
      // };
      // instance.onEnd = () => {
      //   expect(ri).toBe(index);
      // };
      instance.setIndex(index);
    };

    const testIndexList = [0, 1, 2, 3, 4, 5];

    testIndexList.sort(() => Math.random() - 0.5);
    testIndexList.forEach((i) => {
      testIndex(i);
    });

    testIndexList.sort(() => Math.random() - 0.5);
    testIndexList.forEach((i) => {
      testIndex(i);
    });
  });
});
