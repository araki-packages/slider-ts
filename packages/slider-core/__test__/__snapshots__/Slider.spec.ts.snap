// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`increments value on click ループのテスト 1`] = `
"[
  {
    \\"x\\": 305,
    \\"i\\": 1
  },
  {
    \\"x\\": 605,
    \\"i\\": 4
  },
  {
    \\"x\\": 605,
    \\"i\\": 4
  },
  {
    \\"x\\": 539.8148940935188,
    \\"i\\": 3
  },
  {
    \\"x\\": 474.6330570972017,
    \\"i\\": 2
  },
  {
    \\"x\\": 409.45775775728316,
    \\"i\\": 2
  },
  {
    \\"x\\": 344.29226449214644,
    \\"i\\": 1
  },
  {
    \\"x\\": 279.1398452284187,
    \\"i\\": 0
  },
  {
    \\"x\\": 214.00376723709076,
    \\"i\\": 0
  },
  {
    \\"x\\": 648.8872969696688,
    \\"i\\": 4
  },
  {
    \\"x\\": 583.7936998943694,
    \\"i\\": 3
  },
  {
    \\"x\\": 518.7262403323614,
    \\"i\\": 3
  },
  {
    \\"x\\": 453.688181294066,
    \\"i\\": 2
  },
  {
    \\"x\\": 388.6827843155246,
    \\"i\\": 1
  },
  {
    \\"x\\": 323.71330929483725,
    \\"i\\": 1
  },
  {
    \\"x\\": 258.7830143286857,
    \\"i\\": 0
  },
  {
    \\"x\\": 693.8951555489459,
    \\"i\\": 4
  },
  {
    \\"x\\": 629.0529869593986,
    \\"i\\": 4
  },
  {
    \\"x\\": 564.2597602725484,
    \\"i\\": 3
  },
  {
    \\"x\\": 499.51872474655534,
    \\"i\\": 2
  },
  {
    \\"x\\": 434.83312702229136,
    \\"i\\": 2
  },
  {
    \\"x\\": 370.2062109605265,
    \\"i\\": 1
  },
  {
    \\"x\\": 305.64121747925697,
    \\"i\\": 1
  },
  {
    \\"x\\": 241.141384391178,
    \\"i\\": 0
  },
  {
    \\"x\\": 676.7099462413134,
    \\"i\\": 4
  },
  {
    \\"x\\": 612.3501341448107,
    \\"i\\": 4
  },
  {
    \\"x\\": 548.0651756249047,
    \\"i\\": 3
  },
  {
    \\"x\\": 483.85829445106583,
    \\"i\\": 2
  },
  {
    \\"x\\": 419.7327104773326,
    \\"i\\": 2
  },
  {
    \\"x\\": 355.69163948084065,
    \\"i\\": 1
  },
  {
    \\"x\\": 291.73829300056036,
    \\"i\\": 0
  },
  {
    \\"x\\": 227.87587817624444,
    \\"i\\": 0
  },
  {
    \\"x\\": 664.1075975875915,
    \\"i\\": 4
  },
  {
    \\"x\\": 600.4366490936477,
    \\"i\\": 4
  },
  {
    \\"x\\": 536.8662256724374,
    \\"i\\": 3
  },
  {
    \\"x\\": 473.3995152608427,
    \\"i\\": 2
  },
  {
    \\"x\\": 410.0397005947334,
    \\"i\\": 2
  },
  {
    \\"x\\": 346.7899590493589,
    \\"i\\": 1
  },
  {
    \\"x\\": 283.65346248000924,
    \\"i\\": 0
  },
  {
    \\"x\\": 220.6333770629535,
    \\"i\\": 0
  },
  {
    \\"x\\": 657.7328631366595,
    \\"i\\": 4
  },
  {
    \\"x\\": 594.9550750433109,
    \\"i\\": 3
  },
  {
    \\"x\\": 532.3031609706222,
    \\"i\\": 3
  },
  {
    \\"x\\": 469.7802627939627,
    \\"i\\": 2
  },
  {
    \\"x\\": 407.3895159187987,
    \\"i\\": 2
  },
  {
    \\"x\\": 345.13404912345686,
    \\"i\\": 1
  },
  {
    \\"x\\": 283.01698440222344,
    \\"i\\": 0
  },
  {
    \\"x\\": 221.04143680878178,
    \\"i\\": 0
  },
  {
    \\"x\\": 659.2105142999967,
    \\"i\\": 4
  },
  {
    \\"x\\": 597.5273175800598,
    \\"i\\": 3
  },
  {
    \\"x\\": 535.9949399449911,
    \\"i\\": 3
  },
  {
    \\"x\\": 474.6164671275192,
    \\"i\\": 2
  },
  {
    \\"x\\": 413.3949771423354,
    \\"i\\": 2
  },
  {
    \\"x\\": 352.3335401317381,
    \\"i\\": 1
  },
  {
    \\"x\\": 291.4352182116713,
    \\"i\\": 0
  },
  {
    \\"x\\": 230.70306531816368,
    \\"i\\": 0
  },
  {
    \\"x\\": 670.1401270541814,
    \\"i\\": 4
  },
  {
    \\"x\\": 609.7494405368961,
    \\"i\\": 4
  },
  {
    \\"x\\": 549.5340342453783,
    \\"i\\": 3
  },
  {
    \\"x\\": 489.49692786872674,
    \\"i\\": 2
  },
  {
    \\"x\\": 429.6411321546343,
    \\"i\\": 2
  },
  {
    \\"x\\": 369.9696487584056,
    \\"i\\": 1
  },
  {
    \\"x\\": 310.485470092432,
    \\"i\\": 1
  },
  {
    \\"x\\": 251.1915791761262,
    \\"i\\": 0
  },
  {
    \\"x\\": 692.0909494863254,
    \\"i\\": 4
  },
  {
    \\"x\\": 633.186544808183,
    \\"i\\": 4
  },
  {
    \\"x\\": 574.4813190865402,
    \\"i\\": 3
  },
  {
    \\"x\\": 515.9782162777838,
    \\"i\\": 3
  },
  {
    \\"x\\": 457.68017020222214,
    \\"i\\": 2
  },
  {
    \\"x\\": 399.5901043969534,
    \\"i\\": 1
  },
  {
    \\"x\\": 341.7109319692572,
    \\"i\\": 1
  },
  {
    \\"x\\": 284.045555450507,
    \\"i\\": 0
  },
  {
    \\"x\\": 226.596866650616,
    \\"i\\": 0
  },
  {
    \\"x\\": 669.3677465130177,
    \\"i\\": 4
  },
  {
    \\"x\\": 612.3610649701877,
    \\"i\\": 4
  },
  {
    \\"x\\": 555.5796807997294,
    \\"i\\": 3
  },
  {
    \\"x\\": 499.02644148100353,
    \\"i\\": 2
  },
  {
    \\"x\\": 442.70418305234216,
    \\"i\\": 2
  },
  {
    \\"x\\": 386.6157299688148,
    \\"i\\": 1
  },
  {
    \\"x\\": 330.7638949605989,
    \\"i\\": 1
  },
  {
    \\"x\\": 275.1514788919194,
    \\"i\\": 0
  },
  {
    \\"x\\": 219.78127062059139,
    \\"i\\": 0
  },
  {
    \\"x\\": 664.6560468581702,
    \\"i\\": 4
  },
  {
    \\"x\\": 609.7785720306938,
    \\"i\\": 4
  },
  {
    \\"x\\": 555.1515981400617,
    \\"i\\": 3
  },
  {
    \\"x\\": 500.7778646260249,
    \\"i\\": 3
  },
  {
    \\"x\\": 446.6600982288046,
    \\"i\\": 2
  },
  {
    \\"x\\": 392.8010128523547,
    \\"i\\": 1
  },
  {
    \\"x\\": 339.2033094282615,
    \\"i\\": 1
  },
  {
    \\"x\\": 285.8696757803036,
    \\"i\\": 0
  },
  {
    \\"x\\": 232.80278648965395,
    \\"i\\": 0
  },
  {
    \\"x\\": 680.005302760761,
    \\"i\\": 4
  },
  {
    \\"x\\": 627.4798722878913,
    \\"i\\": 4
  },
  {
    \\"x\\": 575.2291291223546,
    \\"i\\": 3
  },
  {
    \\"x\\": 523.2556935404082,
    \\"i\\": 3
  },
  {
    \\"x\\": 471.5621719118608,
    \\"i\\": 2
  },
  {
    \\"x\\": 420.15115656935905,
    \\"i\\": 2
  },
  {
    \\"x\\": 369.02522567839424,
    \\"i\\": 1
  },
  {
    \\"x\\": 318.1869431080122,
    \\"i\\": 1
  },
  {
    \\"x\\": 267.6388583022326,
    \\"i\\": 0
  },
  {
    \\"x\\": 217.38350615220952,
    \\"i\\": 0
  },
  {
    \\"x\\": 667.4234068691048,
    \\"i\\": 4
  },
  {
    \\"x\\": 617.7610658577023,
    \\"i\\": 4
  },
  {
    \\"x\\": 568.3989735907735,
    \\"i\\": 3
  },
  {
    \\"x\\": 519.3396054841814,
    \\"i\\": 3
  },
  {
    \\"x\\": 470.58542177274194,
    \\"i\\": 2
  },
  {
    \\"x\\": 422.1388673868478,
    \\"i\\": 2
  },
  {
    \\"x\\": 374.0023718298662,
    \\"i\\": 1
  },
  {
    \\"x\\": 326.17834905629024,
    \\"i\\": 1
  },
  {
    \\"x\\": 278.6691973506995,
    \\"i\\": 0
  },
  {
    \\"x\\": 231.47729920748043,
    \\"i\\": 0
  },
  {
    \\"x\\": 684.6050212113505,
    \\"i\\": 4
  },
  {
    \\"x\\": 638.0547139186801,
    \\"i\\": 4
  },
  {
    \\"x\\": 591.8287117396167,
    \\"i\\": 3
  },
  {
    \\"x\\": 545.9293328210179,
    \\"i\\": 3
  },
  {
    \\"x\\": 500.35887893019753,
    \\"i\\": 3
  },
  {
    \\"x\\": 455.1196353395053,
    \\"i\\": 2
  },
  {
    \\"x\\": 410.21387071171193,
    \\"i\\": 2
  },
  {
    \\"x\\": 365.6438369862508,
    \\"i\\": 1
  },
  {
    \\"x\\": 321.41176926628395,
    \\"i\\": 1
  },
  {
    \\"x\\": 277.51988570661433,
    \\"i\\": 0
  },
  {
    \\"x\\": 233.97038740244807,
    \\"i\\": 0
  },
  {
    \\"x\\": 690.7654582790228,
    \\"i\\": 4
  },
  {
    \\"x\\": 647.9072649820755,
    \\"i\\": 4
  },
  {
    \\"x\\": 605.3979567691977,
    \\"i\\": 4
  },
  {
    \\"x\\": 563.2396654020476,
    \\"i\\": 3
  },
  {
    \\"x\\": 521.4345050394604,
    \\"i\\": 3
  },
  {
    \\"x\\": 479.9845721314068,
    \\"i\\": 2
  },
  {
    \\"x\\": 438.8919453138733,
    \\"i\\": 2
  },
  {
    \\"x\\": 398.15868530462285,
    \\"i\\": 1
  },
  {
    \\"x\\": 357.7868347998474,
    \\"i\\": 1
  },
  {
    \\"x\\": 317.778418371734,
    \\"i\\": 1
  },
  {
    \\"x\\": 278.13544236693724,
    \\"i\\": 0
  },
  {
    \\"x\\": 238.85989480596254,
    \\"i\\": 0
  },
  {
    \\"x\\": 699.9537452834711,
    \\"i\\": 4
  },
  {
    \\"x\\": 661.4189448695088,
    \\"i\\": 4
  },
  {
    \\"x\\": 623.2574260116671,
    \\"i\\": 4
  },
  {
    \\"x\\": 585.4711024381668,
    \\"i\\": 3
  },
  {
    \\"x\\": 548.0618690618949,
    \\"i\\": 3
  },
  {
    \\"x\\": 511.0316018853755,
    \\"i\\": 3
  },
  {
    \\"x\\": 474.38215790669256,
    \\"i\\": 2
  },
  {
    \\"x\\": 438.11537502636565,
    \\"i\\": 2
  },
  {
    \\"x\\": 402.23307195517555,
    \\"i\\": 2
  },
  {
    \\"x\\": 366.7370481229764,
    \\"i\\": 1
  },
  {
    \\"x\\": 331.62908358843833,
    \\"i\\": 1
  },
  {
    \\"x\\": 296.91093894979076,
    \\"i\\": 0
  },
  {
    \\"x\\": 262.58435525653204,
    \\"i\\": 0
  },
  {
    \\"x\\": 228.65105392211535,
    \\"i\\": 0
  },
  {
    \\"x\\": 695.1127366376322,
    \\"i\\": 4
  },
  {
    \\"x\\": 661.9710852864537,
    \\"i\\": 4
  },
  {
    \\"x\\": 629.2277618599164,
    \\"i\\": 4
  },
  {
    \\"x\\": 596.8844083739586,
    \\"i\\": 3
  },
  {
    \\"x\\": 564.9426467867752,
    \\"i\\": 3
  },
  {
    \\"x\\": 533.4040789174915,
    \\"i\\": 3
  },
  {
    \\"x\\": 502.27028636582236,
    \\"i\\": 3
  },
  {
    \\"x\\": 471.542830432767,
    \\"i\\": 2
  },
  {
    \\"x\\": 441.22325204230765,
    \\"i\\": 2
  },
  {
    \\"x\\": 411.313071664139,
    \\"i\\": 2
  },
  {
    \\"x\\": 381.8137892374143,
    \\"i\\": 1
  },
  {
    \\"x\\": 352.7268840955312,
    \\"i\\": 1
  },
  {
    \\"x\\": 324.05381489194406,
    \\"i\\": 1
  },
  {
    \\"x\\": 295.7960195270116,
    \\"i\\": 0
  },
  {
    \\"x\\": 267.9549150758976,
    \\"i\\": 0
  },
  {
    \\"x\\": 240.53189771750112,
    \\"i\\": 0
  },
  {
    \\"x\\": 213.52834266443824,
    \\"i\\": 0
  },
  {
    \\"x\\": 686.945604094084,
    \\"i\\": 4
  },
  {
    \\"x\\": 660.7850150806644,
    \\"i\\": 4
  },
  {
    \\"x\\": 635.047887528388,
    \\"i\\": 4
  },
  {
    \\"x\\": 609.7355121056826,
    \\"i\\": 4
  },
  {
    \\"x\\": 584.8491581804537,
    \\"i\\": 3
  },
  {
    \\"x\\": 560.390073756429,
    \\"i\\": 3
  },
  {
    \\"x\\": 536.3594854105795,
    \\"i\\": 3
  },
  {
    \\"x\\": 512.7585982316032,
    \\"i\\": 3
  },
  {
    \\"x\\": 489.5885957594983,
    \\"i\\": 2
  },
  {
    \\"x\\": 466.85063992620417,
    \\"i\\": 2
  },
  {
    \\"x\\": 444.54587099733544,
    \\"i\\": 2
  },
  {
    \\"x\\": 422.6754075150038,
    \\"i\\": 2
  },
  {
    \\"x\\": 401.2403462417151,
    \\"i\\": 2
  },
  {
    \\"x\\": 380.2417621053846,
    \\"i\\": 1
  },
  {
    \\"x\\": 359.6807081454135,
    \\"i\\": 1
  },
  {
    \\"x\\": 339.5582154598942,
    \\"i\\": 1
  },
  {
    \\"x\\": 319.8752931538984,
    \\"i\\": 1
  },
  {
    \\"x\\": 300.63292828886733,
    \\"i\\": 1
  },
  {
    \\"x\\": 281.8320858331226,
    \\"i\\": 0
  },
  {
    \\"x\\": 263.4737086134646,
    \\"i\\": 0
  },
  {
    \\"x\\": 245.55871726790247,
    \\"i\\": 0
  },
  {
    \\"x\\": 228.08801019947168,
    \\"i\\": 0
  },
  {
    \\"x\\": 211.06246353119786,
    \\"i\\": 0
  },
  {
    \\"x\\": 694.4829310621444,
    \\"i\\": 4
  },
  {
    \\"x\\": 678.3502442246063,
    \\"i\\": 4
  },
  {
    \\"x\\": 662.6652120424133,
    \\"i\\": 4
  },
  {
    \\"x\\": 647.4286210903574,
    \\"i\\": 4
  },
  {
    \\"x\\": 632.6412354547483,
    \\"i\\": 4
  },
  {
    \\"x\\": 618.303796695096,
    \\"i\\": 4
  },
  {
    \\"x\\": 604.4170238069237,
    \\"i\\": 4
  },
  {
    \\"x\\": 590.9816131857115,
    \\"i\\": 3
  },
  {
    \\"x\\": 577.9982385919684,
    \\"i\\": 3
  },
  {
    \\"x\\": 565.4675511174537,
    \\"i\\": 3
  },
  {
    \\"x\\": 553.3901791525186,
    \\"i\\": 3
  },
  {
    \\"x\\": 541.7667283545998,
    \\"i\\": 3
  },
  {
    \\"x\\": 530.5977816178365,
    \\"i\\": 3
  },
  {
    \\"x\\": 519.8838990438526,
    \\"i\\": 3
  },
  {
    \\"x\\": 509.6256179136599,
    \\"i\\": 3
  },
  {
    \\"x\\": 499.8234526607139,
    \\"i\\": 2
  },
  {
    \\"x\\": 490.4778948451258,
    \\"i\\": 2
  },
  {
    \\"x\\": 481.58941312899515,
    \\"i\\": 2
  },
  {
    \\"x\\": 473.15845325292685,
    \\"i\\": 2
  },
  {
    \\"x\\": 465.18543801366286,
    \\"i\\": 2
  },
  {
    \\"x\\": 457.6707672428802,
    \\"i\\": 2
  },
  {
    \\"x\\": 450.6148177871564,
    \\"i\\": 2
  },
  {
    \\"x\\": 444.0179434890506,
    \\"i\\": 2
  },
  {
    \\"x\\": 437.8804751693715,
    \\"i\\": 2
  },
  {
    \\"x\\": 432.20272061058495,
    \\"i\\": 2
  },
  {
    \\"x\\": 426.98496454137785,
    \\"i\\": 2
  },
  {
    \\"x\\": 422.2274686223773,
    \\"i\\": 2
  },
  {
    \\"x\\": 417.9304714330374,
    \\"i\\": 2
  },
  {
    \\"x\\": 414.0941884596632,
    \\"i\\": 2
  },
  {
    \\"x\\": 410.7188120846149,
    \\"i\\": 2
  },
  {
    \\"x\\": 407.80451157665084,
    \\"i\\": 2
  },
  {
    \\"x\\": 405.35143308245097,
    \\"i\\": 2
  },
  {
    \\"x\\": 403.3596996192755,
    \\"i\\": 2
  },
  {
    \\"x\\": 401.82941106880025,
    \\"i\\": 2
  },
  {
    \\"x\\": 400.76064417211455,
    \\"i\\": 2
  },
  {
    \\"x\\": 400.1534525258612,
    \\"i\\": 2
  },
  {
    \\"x\\": 400.0078665795554,
    \\"i\\": 2
  },
  {
    \\"x\\": 400,
    \\"i\\": 2
  }
]"
`;
